pkg/server/image_pull.go

```diff
@@ -17,14 +17,202 @@ limitations under the License.
+			}
+		}
+	}
+
+	// Return the image digest
+	return &runtime.PullImageResponse{ImageRef: digest}, err
+}
+
+// imageReferenceResolver attempts to resolve the image reference into a name
+// and manifest via the containerd library call..
+//
+// The argument `ref` should be a scheme-less URI representing the remote.
+// Structurally, it has a host and path. The "host" can be used to directly
+// reference a specific host or be matched against a specific handler.
+//
+// The returned name should be used to identify the referenced entity.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+//
+// If the resolution fails, an error will be returned.
+func (c *criContainerdService) imageReferenceResolver(ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
+	var (
+		ctx  = gocontext.Background()
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler to fetch the object the manifest
+	resolver := docker.NewResolver()
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		var (
+			err     error
+			fetcher remotes.Fetcher
+		)
+
+		resolvedImageName, desc, fetcher, err = resolver.Resolve(ctx, ref)
+		if err != nil {
+			return err
+		}
+
+		eg.Go(func() error {
+			return c.imageStore.Put(ctx, resolvedImageName, desc)
+		})
+
+		return containerd_images.Dispatch(
+			ctx,
+			remotes.FetchHandler(c.ingester, fetcher),
+			desc)
+	})
+
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		errs <- eg.Wait()
+	}()
+
+	defer func() {
+		nctx := gocontext.Background()
+		image, err := c.imageStore.Get(nctx, resolvedImageName)
+		if err != nil {
+			fmt.Printf("\nc.imageStore.Get for %q ERR: %v\n", resolvedImageName, err)
+		}
+		p, err := content.ReadBlob(nctx, c.provider, image.Target.Digest)
+		if err != nil {
+			fmt.Printf("\ncontent.ReadBlob ERR: %v\n", err)
+		}
+		err = json.Unmarshal(p, &manifest)
+		if err != nil {
+			fmt.Printf("\njson.Unmarshal ERR: %v\n", err)
+		}
+		size, err = image.Size(nctx, c.provider)
+		if err != nil {
+			fmt.Printf("\nimage.Size ERR: %v\n", err)
+		}
+		compressedSize = uint64(size)
+	}()
+
+	select {
+	case err := <-errs:
+		return resolvedImageName, manifest, compressedSize, err
+	case <-ctx.Done():
+		return resolvedImageName, manifest, compressedSize, nil
+	}
+}
+
+func (c *criContainerdService) pullImage(ref string) (imagespec.Descriptor, int64, error) {
+	var (
+		ctx               = gocontext.Background()
+		size              int64
+		desc              imagespec.Descriptor
+		resolvedImageName string
+	)
+
+	// This first go routine in the local error group eg will resolve the image
```

Comment:
dist use goroutine because they need to report progress.

For us, just call the functions one by one is good enough.